### PR TITLE
enchantify special vision [1/2]

### LINKED ISSUE #71593: JSONify enchantment-based detection
### Is your feature request related to a problem? Please describe.

First, I thought "Oh, that motion sensor is neat! I should make a telepathic power that lets you sense other minds like I've been thinking about." And I went through and edited the .cpp and .h files and did it, and it all worked.

Then I thought, "Oh, for Xedra Evolved I should make a power that lets you sense other fair folk, that is thematic and cool." And I went through and edited the .cpp and .h files and did it, and it all worked.

Then I thought, "Oh, for Magiclysm I should-" and then I stopped because editing a ton of niche enchants into the code seems like it will very rapidly become untenable.

### Solution you would like.

Some kind of JSON object that a DETECT_CREATURE enchant would read, looking for NPC/not NPC, creature species, monster flag, npc_has_effect, only NPCs wielding weapons with the DURABLE_MELEE flag, whatever the parameters are, and then the enchant-bearer would be able to sense those creatures and those creatures only. 

### Describe alternatives you have considered.

Hardcoding everything. I only considered this for a moment before rejecting it.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by GuardianDll**
#### Summary
None
#### Purpose of change
Resolve #71593
#### Describe the solution
Made it json
#### Testing
Works pretty well, but actual replacement of enchantment values gonna happen after this one, it would be pretty massive json movements, and i want to solve few todos and maybe polish a code a bit - i do not like that i need to rerun entire function every here and there
image
image

### CODE REVIEW

File: data/json/enchantments.json [hunk 1]
```diff
@@ -43,6 +43,128 @@
...
+          },
+          {
+            "id": "infrared_creature_small",
+            "text_condition": { "math": [ "n_val('size') == 2" ] },
+            "text": "You see a small figure radiating heat."
+          },
+          {
+            "id": "infrared_creature_medium",
+            "text_condition": { "math": [ "n_val('size') == 3" ] },
+            "text": "You see a medium figure radiating heat"
```
Comment (Night-Pryanik) [inline data/json/enchantments.json:69]:
```suggestion
            "text": "You see a medium figure radiating heat."
```

File: data/json/enchantments.json [hunk 2]
```diff
@@ -43,6 +43,128 @@
...
+            "text_condition": { "math": [ "n_val('size') == 4" ] },
+            "text": "You see a large figure radiating heat."
+          },
+          {
+            "id": "infrared_creature_huge",
+            "text_condition": { "math": [ "n_val('size') == 5" ] },
+            "text": "You see a huge figure radiating heat."
+          },
+          {
+            "id": "infrared_creature_medium",
```
Comment (Night-Pryanik) [inline data/json/enchantments.json:82]: The description with this id is already defined up above?

File: data/json/enchantments.json [hunk 3]
```diff
@@ -43,6 +43,128 @@
...
+          },
+          {
+            "id": "infrared_creature_small",
+            "text_condition": { "math": [ "n_val('size') == 2" ] },
+            "text": "You see a small figure radiating heat."
+          },
+          {
+            "id": "infrared_creature_medium",
+            "text_condition": { "math": [ "n_val('size') == 3" ] },
+            "text": "You see a medium figure radiating heat"
```
Comment (Night-Pryanik) [inline data/json/enchantments.json:112]:
```suggestion
            "text": "You see a medium figure radiating heat."
```

File: data/json/enchantments.json [hunk 4]
```diff
@@ -43,6 +43,128 @@
...
+  {
+    "type": "enchantment",
+    "id": "THERMAL_VISION_LEAFNOSE",
+    "condition": {
+      "not": {
+        "or": [
+          { "u_has_effect": "smoke", "bodypart": "mouth" },
+          { "u_has_effect": "asthma" },
+          { "u_has_effect": "tpollen" },
+          { "u_has_effect": "asthma" },
```
Comment (Night-Pryanik) [inline data/json/enchantments.json:142]: Double check for asthma.
```suggestion
```

File: doc/MAGIC.md [hunk 5]
```diff
@@ -666,6 +666,45 @@ There are two possible syntaxes.  The first is by defining an enchantment object
...
     }
+    "special_vision": [   // defines creatures (monsters or NPC) you can see in some irregular ways, mainly thermal or supernatural
+      {
+        "condition": { "and": [ { "npc_has_flag": "ELECTRIC" }, "u_see_npc" ] },   // this need to return true to see the critter; u is character, npc is critter being watched
+        "distance": 20, // how far special_vision is applied. for technical reasons special_vision do not stack, having multiple special_visions of the same nature will not result in special_visions of bigger distance
+        "descriptions": [ // if condition is true, this will assign a dedicated id (used for tiles) and text, depending on text_condition
+          { 
+            "id": "infrared_creature_tiny", 
+            "text_condition": { "math": [ "n_val('size') == 1" ] }, // be sure to not use condition, that can change in between moves (like any math with random result, `rand(1)` etc); 
+                        // while text with tags is stored and re-evaluated only when cursor chages it's position, tiles are re-evaluated every frame, even on pause
```
Comment (Night-Pryanik) [inline doc/MAGIC.md:677]:
```suggestion
            "text_condition": { "math": [ "n_val('size') == 1" ] }, // be sure to not use condition that can change in between moves (like any math with random result, `rand(1)` etc); 
                        // while text with tags is stored and re-evaluated only when cursor changes its position, tiles are re-evaluated every frame, even on pause
```

File: data/json/enchantments.json [hunk 6]
```diff
@@ -43,6 +43,128 @@
...
+          },
+          {
+            "id": "infrared_creature_huge",
+            "text_condition": { "math": [ "n_val('size') == 5" ] },
+            "text": "You see a huge figure radiating heat."
+          },
+          {
+            "id": "infrared_creature_medium",
+            "text_condition": { "math": [ "true" ] },
+            "text": "You see a figure radiating heat"
```
Comment (Night-Pryanik) [inline data/json/enchantments.json:84]:
```suggestion
            "text": "You see a figure radiating heat."
```

File: data/json/enchantments.json [hunk 7]
```diff
@@ -43,6 +43,128 @@
...
```
Comment (Night-Pryanik) [inline data/json/enchantments.json:127]:
```suggestion
            "text": "You see a figure radiating heat."
```

**Comment by GuardianDll** [comment]
There is a lot of duplicate code in enchantments overall, and i am not familiar with the ways to simiplify it, except templates (which i do not think is fitting here)

**Pull request merged**